Finish refactoring subscriber related methods in the Sending class [MAILPOET-4368]

rodrigoprimo opened the pull request:
## Description

In https://github.com/mailpoet/mailpoet/pull/5018, we refactored two subscriber related methods in the `Sending` class to use Doctrine instead of Paris. This refactors the remaining subscriber related methods. It also removes `mailpoet/lib/Tasks/Subscribers.php`.

Historically, we have used an incremental approach to refactor Paris to Doctrine. I'm a bit worried about it in the case of the Sending class as I had to fix some bugs that would have been easy to miss if not for our automated tests. All the bugs were related to Doctrine entities having stale data due to changes made using Paris. There are still several parts of the Sending class that use Paris so this PR might introduce more bugs that were not caught by our automation tests. I tried to add as many tests as possible, but the problem is that those bugs don't manifest themselves directly in the parts of the code that were changed.

That being said, I'm not sure if refactoring the whole Sending class in one push is feasible, but maybe we can build upon this PR and try it?

## Code review notes

As @lysyjan pointed out in this comment https://github.com/mailpoet/mailpoet/pull/5018#issuecomment-1638302275, I ended up taking an approach that is different than what was initially thought when the ticket was created. The idea was to move the subscriber related methods out of the Sending class maybe to repository classes. I refactored the methods to use Doctrine but kept them in the Sending class. We need to decide if we want to create a separate ticket to move the method out of the Sending class or if we tackle this in https://mailpoet.atlassian.net/browse/MAILPOET-4375. I'm inclined to think the first approach is the best one but I'm open to suggestions.

This PR was created on top of https://github.com/mailpoet/mailpoet/pull/5018, so the first six commits belong to https://github.com/mailpoet/mailpoet/pull/5018.

## QA notes

This PR should not change any functionality, but it touches sending emails which is a core part of MailPoet. It would be nice to test anything sending related that is not covered by automated tests.

## Linked PRs

https://github.com/mailpoet/mailpoet/pull/5018

## Linked tickets

[MAILPOET-4368]

## After-merge notes

_N/A_

## Tasks

- [X] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [X] I added sufficient test coverage
- [X] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes

[MAILPOET-4368]: https://mailpoet.atlassian.net/browse/MAILPOET-4368?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

rodrigoprimo committed:
Improve test coverage for Sending::removeSubscribers()

[MAILPOET-4368]

rodrigoprimo committed:
Fix typo

[MAILPOET-4368]

rodrigoprimo committed:
Update checkCompleted() to also change the processedAt field

This commit updates the method ScheduledTaskSubscribersRepository::checkCompleted() to update the processedAt field of the Scheduled Task. The method was added in f62d9b416766e4fdff60ff6dd022da545075c119 and I believe it is meant to be a replacement to \MailPoet\Tasks\Subscribers::checkCompleted(). The original method updates the processedAt field, so I'm assuming the new method should update as well. I found this while working on refactoring \MailPoet\Tasks\Sending::removeSubscribers() to use Doctrine as this method calls \MailPoet\Tasks\Subscribers::checkCompleted() internaly.

[MAILPOET-4368]

rodrigoprimo committed:
Refactor Sending::removeSubscribers() to use Doctrine instead of Paris

[MAILPOET-4368]

rodrigoprimo committed:
Remove unused method removeAllSubscribers()

This method was added in c0c57f6b67ba5a26ffbf680fdb9b59b83ace9dad but
its only usage besides integration test was removed in
18a6a1c439735ba6e56ae163f9fbd3d6cda9bcd8. It doesn't seem like it is
still used in our code base.

[MAILPOET-4368]

rodrigoprimo committed:
Update test coverage for updateProcessedSubscribers()

[MAILPOET-4368]

rodrigoprimo committed:
Add the ScheduledTaskEntity as a property to the Sending class

Doing this instead of having to get the ScheduledTaskEntity in several
methods of this class.

[MAILPOET-4368]

rodrigoprimo committed:
Refactor updateProcessedSubscribers() to use Doctrine instead of Paris

It was necessary to update a test to be more lenient when comparing
dates and another one to persist a change in Paris to the database.

[MAILPOET-4368]

rodrigoprimo committed:
Refactor Sending::delete() to use Doctrine instead of Paris

[MAILPOET-4368]

rodrigoprimo committed:
Add integration test for Sending::saveSubscriberError()

[MAILPOET-4368]

rodrigoprimo committed:
Refactor Sending::saveSubscriberError() to use Doctrine instead of Paris

[MAILPOET-4368]

rodrigoprimo committed:
Remove \MailPoet\Tasks\Subscribers class now that it is not used anymore

[MAILPOET-4368]

rodrigoprimo committed:
Remove remaining uses of the ScheduledTaskSubscriber model from Sending

[MAILPOET-4368]

rodrigoprimo committed:
Replace Newsletter model with NewsletterEntity in SendingQueue

This was the easiest way to fix failing integration tests related to
updating some Sending methods to use Doctrine instead of Paris. Using
Paris in SendingQueue to update the database was creating a state where
Doctrine had stale data and this was causing the tests to fail.

[MAILPOET-4368]

rodrigoprimo committed:
Refresh Doctrine entities in SendingQueue::add() to avoid issues

Since SendingQueue::add() still uses Paris, it was running into some
problems caught by our integration tests since some Sending methods
where refactored to Doctrine. The issue was that we started loading some
Doctrine entities earlier, then Paris code inside SendingQueue::add()
would change the same rows, causing the Doctrine entities to contain
stale data. To workaround this problem, in this commit, we are making
sure that the Doctrine entities are refreshed after they are changed by
Paris.

I considered refactoring SendingQueue::add() to use Doctrine but I
wasn't sure what to do with SendingTask::create() and
SendingTask::createFromQueue() so I decided to take a different
approach.

[MAILPOET-4368]

JanJakes commented on mailpoet/lib/Newsletter/Sending/ScheduledTaskSubscribersRepository.php:
```diff
@@ -140,6 +141,7 @@ private function checkCompleted(ScheduledTaskEntity $task): void {
     $count = $this->countBy(['task' => $task, 'processed' => ScheduledTaskSubscriberEntity::STATUS_UNPROCESSED]);
     if ($count === 0) {
       $task->setStatus(ScheduledTaskEntity::STATUS_COMPLETED);
+      $task->setProcessedAt(new Carbon());
```
Nice catch!

JanJakes commented on mailpoet/lib/Tasks/Sending.php:
```diff
@@ -276,8 +276,10 @@ public function removeSubscribers(array $subscriberIds) {
-    $this->taskSubscribers->updateProcessedSubscribers($processedSubscribers);
+  public function updateProcessedSubscribers(array $processedSubscribers): bool {
+    $this->scheduledTaskSubscribersRepository->updateProcessedSubscribers($this->scheduledTaskEntity, $processedSubscribers);
+    $this->scheduledTasksRepository->refresh($this->scheduledTaskEntity); // needed while Sending still uses Paris
```
I wonder if that comment is accurate. The change was done using Doctrine, so the refresh is probably not needed for Paris, but rather due to the fact that some query didn't reflect changes to a Doctrine entity?

That said, is the refresh needed at all? It seems that `updateProcessedSubscribers` and `checkCompleted` don't use DBAL, so the changes on `$this->scheduledTaskEntity` could simply be up-to-date.

rodrigoprimo replied:
> I wonder if that comment is accurate. The change was done using Doctrine, so the refresh is probably not needed for Paris, but rather due to the fact that some query didn't reflect changes to a Doctrine entity?

Maybe I could rephrase the comment (please let me know if you think that is necessary and if you have a suggestion), but I'm not trying to say that the refresh is needed for Paris, but rather, while we still use Paris in the Sending class.

> That said, is the refresh needed at all? It seems that updateProcessedSubscribers and checkCompleted don't use DBAL, so the changes on $this->scheduledTaskEntity could simply be up-to-date.

This refresh is still needed. Without it, the test `\MailPoet\Test\Cron\Workers\SendingQueue\SendingQueueTest::testItDoesNotEnforceExecutionLimitsAfterSendingWhenQueueStatusIsSetToComplete()` and maybe others (I haven't checked while writing this comment) fail. This is the tricky thing about the changes introduced in this PR, they have the potential to break seemingly unrelated parts of the code. Speaking off the top of my head, using Doctrine here means that we are loading the Doctrine entity earlier. Later in the code execution, Paris code changes the rows associated with the loaded Doctrine entities and they became outdated. Calling refresh here, forces Doctrine to query the database again the next time the entity is used.

JanJakes replied:
@rodrigoprimo Ahh, I understand now, thanks for the clarification!

JanJakes commented on mailpoet/lib/Tasks/Sending.php:
```diff
@@ -209,22 +213,30 @@ public function save() {
+    $this->scheduledTaskSubscribersRepository->deleteByScheduledTask($this->scheduledTaskEntity);
+    $this->scheduledTasksRepository->remove($this->scheduledTaskEntity);
+
+    $sendingQueueEntity = $this->sendingQueuesRepository->findOneById($this->queue->id);
```
You can also do `$this->scheduledTasksRepository->getSendingQueue()`. It's a tiny improvement, but it eliminates the usage of `$this->queue`.

rodrigoprimo replied:
Thanks for your attention to the details: 4fcdd9a7f0b1bca4e421009a2ef4955fea628374

JanJakes commented on mailpoet/lib/Tasks/Sending.php:
```diff
@@ -209,22 +213,30 @@ public function save() {
+
+    if (!$sendingQueueEntity) {
+      throw new InvalidStateException();
+    }
```
Rather than throwing an exception here, I think we could do this:

```php
if ($sendingQueueEntity) {
  $this->sendingQueuesRepository->remove($sendingQueueEntity);
}
```

That way, if the data is for any reason incorrect, it will still pass.

rodrigoprimo replied:
I'm ok with this change: cf54db89500a8580b77f3ffaed0999e2ad950194

JanJakes commented on mailpoet/lib/Entities/SendingQueueEntity.php:
```diff
@@ -223,4 +223,36 @@ public function getNewsletter() {
     $this->newsletter = $newsletter;
   }
+
+  public function toArray(): array {
```
I think the specific API response format belongs rather to [ResponseBuilders](https://github.com/mailpoet/mailpoet/tree/trunk/mailpoet/lib/API/JSON/ResponseBuilders), e.g. a new `SendingQueueResponseBuilder` class.

rodrigoprimo replied:
I saw a few entities with a `toArray()` method, and that threw me off. I have moved the code to a ResponseBuilder in bdbfdbc7626afdf5cdf2691fc553cf9d640bb06f

JanJakes commented on mailpoet/lib/Entities/SendingQueueEntity.php:
```diff
@@ -223,4 +223,36 @@ public function getNewsletter() {
+
+  public function toArray(): array {
+    if (!$this->getTask() instanceof ScheduledTaskEntity) {
+      throw new \RuntimeException('Invalid state. SendingQueue has ScheduledTask associated.');
```
I think the message is missing a negation, such as "SendingQueue has **no** ScheduledTask associated."

rodrigoprimo replied:
Done in e7ba64c26f2705c9b57023939750b3cbb2684b86

rodrigoprimo committed:
Fix typo in error message

[MAILPOET-4368]

rodrigoprimo committed:
Move SendingQueueEntity::toArray() to SendingQueuesResponseBuilder::build()

Based on a suggestion during the review of the PR, it was decided that
it makes more sense for the code that was created as
SendingQueueEntity::toArray() tp be moved to
SendingQueuesResponseBuilder::build(). It is used only to build the
format that is send in the API response.

[MAILPOET-4368]

rodrigoprimo committed:
Minor improvement: use ScheduledTaskEntity::getSendingQueue()

Using it to avoid using $this->queue->id

[MAILPOET-4368]

rodrigoprimo committed:
Don't thrown an exception if data is incorrect

See https://github.com/mailpoet/mailpoet/pull/5238#discussion_r1371462690

[MAILPOET-4368]

rodrigoprimo commented:
Just documenting that I had to rebase this branch and force push to fix a conflict. No code changes were needed to fix the conflicts and git was able to fix everything on its own.

rodrigoprimo commented:
@JanJakes, I believe I addressed all of your comments. Could you please check this PR again?

JanJakes reviewed (approved):
LGTM, thanks for the fixes!

Aschepikov reviewed (approved):
GJ!
